cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java

```diff
@@ -58,17 +72,41 @@ protected ModelAndView internalHandleRequest(final String method, final HttpServ
         final ServiceTicket serviceTicket = (ServiceTicket) ticketRegistry.getTicket(code);
         // service ticket should be valid
         if (serviceTicket == null || serviceTicket.isExpired()) {
-            logger.error("Code expired : {}", code);
+            logger.error("Code expired: {}", code);
             return OAuthUtils.writeTextError(response, OAuthConstants.INVALID_GRANT, HttpStatus.SC_BAD_REQUEST);
         }
         final TicketGrantingTicket ticketGrantingTicket = serviceTicket.getGrantingTicket();
         // remove service ticket
         ticketRegistry.deleteTicket(serviceTicket.getId());
 
-        final String accessTokenEncoded = this.accessTokenGenerator.generate(serviceTicket.getService(), ticketGrantingTicket);
+        final OAuthRegisteredService service = OAuthUtils.getRegisteredOAuthService(this.servicesManager, clientId);
+        final Principal principal = ticketGrantingTicket.getAuthentication().getPrincipal();
+        final Map<String, Object> attributes = new HashMap<>(service.getAttributeReleasePolicy().getAttributes(principal));
+        final String id = principal.getId();
+        if (!service.getAccessStrategy().doPrincipalAttributesAllowServiceAccess(id, attributes)) {
+            logger.error("Service [{}] is not authorized for use by [{}].", service.getServiceId(), principal);
+            return OAuthUtils.writeTextError(response, OAuthConstants.INVALID_GRANT, HttpStatus.SC_BAD_REQUEST);
```

Comment:
For now, I'd prefer to keep this method "as is", because this is the first PR of the OAuth series, so I'm not sure if it will be needed later on.